From: Edward Adam Davis <eadavis@qq.com>
Subject: copy_process(): fix jump logic error
Date: Wed, 24 Sep 2025 20:06:21 +0800

After futex_hash_allocate_default() fails, the logic should jump to
bad_fork_cancel_cgroup, not bad_fork_core_free.

Jumping to bad_fork_core_free would cause a siglock imbalance.

Link: https://lkml.kernel.org/r/tencent_5E4F978D9525A58D97925ADDBADDB2193107@qq.com
Signed-off-by: Edward Adam Davis <eadavis@qq.com>
Cc: David Hildenbrand <david@redhat.com>
Cc: Kees Cook <kees@kernel.org>
Cc: Oleg Nesterov <oleg@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 kernel/fork.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/kernel/fork.c~copy_process-fixed-jump-logic-error
+++ a/kernel/fork.c
@@ -2293,7 +2293,7 @@ __latent_entropy struct task_struct *cop
 	if (need_futex_hash_allocate_default(clone_flags)) {
 		retval = futex_hash_allocate_default();
 		if (retval)
-			goto bad_fork_core_free;
+			goto bad_fork_cancel_cgroup;
 		/*
 		 * If we fail beyond this point we don't free the allocated
 		 * futex hash map. We assume that another thread will be created
_
